### DISCUSSION TITLE
Clipboard

### DISCUSSION THREAD

**Discussion by JackBouchac** [amidaware/tacticalrmm#1134] [Q&A]
Hello, i would like to ask if ctrl+c and ctrl+v or right click copy/paste works for you in take control. I am using Google Chrome and have checked automatic cliboard in settings but is does not work. If i turn off automatic clipboard, two new buttons show up down right but it does not work through them either. The only way i can send text from my computer is using down left "Type" button and remote keyboard. thank you

**Comment by dinger1986** [comment]
can you give more information about your setup? I think this is a bug in version 1.0.22 of mesh as not working for me

**Reply by JackBouchac** [reply]
I have latest rmm v0.13.4 on dedicated Ubuntu 20.04.4 LTS fully apt, agent v2.0.4, lastest Windows 11 Pro x64, Google Chrome 101.0.4951.67 and Firefox 100.0. Google Chrome and Edge have same behaviour automatic and manual (automatic clipboard unchecked in settings) does not work. In Firefox theres missing "automatic clipboard" in settings and down-right is only one choice "get remote clipboard to local" and this is working, but only with manualy clicked and not with crtl+c and ctrl+v.

**Reply by dinger1986** [reply]
The remote side is meshcentral not tactical, I presume you get the same behaviour if you go to mesh.yourdomain.com and access the machine from there.

**Reply by JackBouchac** [reply]
Tested in Chrome:
mesh.mydomain auto clipboard ctrl+c/v and right click copy/paste from local to remote -> working
mesh.mydomain auto clipboard ctrl+c/v and right click copy/paste from remote to local -> not working
mesh.mydomain manual clipboard ctrl+c/v and right click copy/paste from local to remote -> working
mesh.mydomain manual clipboard ctrl+c/v and right click copy/paste from remote to local -> working
rmm.mydomain -> never worked even with all permission enabled

When i first connected to mesh.mydomain.com and went to remote, Chrome asked me for enable access to clipboard, rmm.mydomain does not asked this. I tried add permission manualy to https://rmm.mydomain.com:443 (exactly same as mesh.mydomain did) and it did not help. Interesting is when i am connected through mesh.mydomain, remote window have third icon on the right part of the address bar and it says "enabled access to read clipboard", "this page have access to text and images copied to clipboard" but remote window through rmm.mydomain does not have this. Can you confirm that copy/paste sometimes worked for you from rmm.mydomain?

**Comment by lumpycustard** [comment]
Having the same issue.

No matter what settings I mess around with TacticalRMM remote clipboard does not work, whether I use automatic clipboard or manually click the send/receive clipboard button in the status bar.

I also tried modifying settings in Chrome/Firefox to allow clipboard (even though TRMM never requests permission anyway).

Using mesh.mydomain.com, remote clipboard works fine. On first visit my browser prompted me for clipboard access.

This bug has persisted for me over the last few months as i evaluate TacticalRMM for enterprise usage. The lack of remote clipboard makes TRMM a bit of a chore to use.

Hopefully this gets some attention soon 👍

**Reply by bbrendon** [reply]
Nobody here really contributes code to mesh. You're probably better of going to their home and asking.

**Reply by lumpycustard** [reply]
Thanks for the reply.

My testing shows clipboard works fine in mesh.yourdomain.com, so that lead to me to believe the issue rests with TRMM since it's presenting it via its front end.

**Reply by wh1te909** [reply]
this might be because of new clipboard blocking features of browsers specifically when in iframes and cross domain. can you open dev tools of the iframe (trmm frontend) and see if there are any permission errors in console? we might need to explicitly add these perms in the iframe element

**Reply by lumpycustard** [reply]
Hi there, yes can confirm errors are thrown up in the console. When I visit mesh.domain.com, the browser asks for clipboard permissions. TRMM does not. Forcing clipboard permissions via site settings in Chrome or Firefox doesn't do anything.

Console output after pressing the "send clipboard" button in the toolbar of TRMM:
![image](https://github.com/amidaware/tacticalrmm/assets/19260650/de35e3a8-f860-4214-a6b3-7c4d5aa9bee8)

Seems this is a known thing with iframes, though I'm not a programmer :)
https://hellodevworld.com/misc/copying-to-clipboard-from-iframe

**Reply by wh1te909** [reply]
@lumpycustard great find thank you. I'll add those permissions to the iframe element and let you know if it works

**Reply by lumpycustard** [reply]
Great news, thank you for the update!

**Reply by wh1te909** [reply]
just FYI I have actually not been able to reproduce this error from my testing. Clipboard is working fine for me in trmm's iframe and I don't have any clipboard policies set that I can find, so I'm not sure what's going on on your end or if there is some policy I have enabled that I'm not aware of (I'm just using regular chrome, tested incognito as well and still works).

**Reply by lumpycustard** [reply]
I have set up TRMM in two separate instances, one for home, and one for testing at work. Both of them have been deployed via Linode following your setup guide to the letter. Certificates are set up, 2FA set up, everything works perfectly fine.

I have tested Chrome, Edge, Vivaldi, Safari on MacOS, all exhibit the exact same behaviour so i'm not sure if i've not set up TRMM properly in both of my Linode VPS instances.

Has happened in Windows 10 22H2, and Windows 11 22H2 as well, on desktops and laptops, as well as on personal machines and domain joined machines.

There are dozens of variables (different computers, browsers, operating systems, desktops and laptops), the only constant is the TRMM instance i am connecting to, so perhaps something was not configured properly during initial setup -- though like i said i followed the guide to the letter and everything is working as expected.

**Comment by Elionis** [comment]
Hi, Same problem for me.
ctrl+c and ctrl+v or right click copy/paste not work on Tactical but work directly on Mesh with only Chome.
Regards

**Comment by tricky1024** [comment]
I have the same issue and again looking at the dev tools i get these errors:

![image](https://github.com/amidaware/tacticalrmm/assets/16222971/ca77d516-a04e-4280-b9a4-ee0c8392857b)

**Reply by tricky1024** [reply]
i also found this:
https://stackoverflow.com/questions/61401384/can-text-within-an-iframe-be-copied-to-clipboard

**Comment by btnalexandre** [comment]
Same problem for me. Please fix this, it's unbearable to manage from a distance

**Comment by lumpycustard** [comment]
Is it possible this is happening as a result of using an unsigned powershell script version of the the mesh client??

**Reply by tricky1024** [reply]
I don't think so as if I run the remote from MESH instead of TRMM it works as expected, it does appear to be an imbedded iframe issue.

**Reply by sqvdotbe** [reply]
Same problem here with TRMM v0.19.0 within Edge, Chrome or Firefox on Win11. Can't copy paste natively, need to use the 'Type' window for pasting something on the remote computer.

**Reply by btnalexandre** [reply]
Yes this is very annoying ...

**Comment by mtsinformatica** [comment]
To fix this issue, you need to add this code to the iframe:
allow="clipboard-read; clipboard-write"
I've tested in Chrome with DevTools and i can copy/past automaticaly in RDP Connection

**Reply by tricky1024** [reply]
can you let me have the details of where you added this and ill post details on the sponsors group so they can take a further look, i would love to get this sorted

**Reply by tricky1024** [reply]
I have been playing with DEVTOOLS and cannot get this to work, even more cannot even get the iframe to acknowledge that clipboard-read and clipboard-write are even permitted under iframe details. #mtsinformatica please provide your details of where you added in DEVTOOLS.

**Reply by btnalexandre** [reply]
It would be nice to be implemented directly in tacticalrmm instead of playing with dev tools 😭

**Reply by tricky1024** [reply]
DEVTOOLS to test only, i can then pass on to the sponsors group for implementation and testing.

**Reply by mtsinformatica** [reply]
![image](https://github.com/user-attachments/assets/9d695361-c4aa-4092-9e6e-868d3f671187)
I've added the code i've posted in the previous post, then i reload the iframe and the browse ask me to allow clipboard

**Reply by tricky1024** [reply]
does that also work for a straight remote or only RDP?

**Reply by mtsinformatica** [reply]
I've only tested with RDP, but it might work with straight remote too, if it's a permissions issue.

**Reply by tricky1024** [reply]
so if i enable like you in the iframe: 

<img width="356" alt="image" src="https://github.com/user-attachments/assets/0b08bbbb-adf3-4d99-89f6-0965d4ec4059" />

i then get prompted to reload the page, i do so and can see the settings remain in devtools but if i then within dev tools click show iframe details i can still see both read and write are disabled:

<img width="425" alt="image" src="https://github.com/user-attachments/assets/fc911cb0-22bb-40ad-bea3-cdd209c42788" />

how are you reloading the iframe?

**Reply by wh1te909** [reply]
thank you, I've added `allow="clipboard-read; clipboard-write"` to the iframe. it will be in the next release

**Comment by wh1te909** [comment]
this issue is now resolved in the latest release 0.20.1

**Reply by DjMagicFingers** [reply]
Not quite yet :)
It works only in one direction (copy at my side, paste on client computer) and it seems like it takes some time and somethimes a 2nd click to the content of the clipboard to actually be present on the remote computer until it is available for "paste" command.
I have tried single short text line - nothing special.

**Reply by tricky1024** [reply]
i have the same, i can copy from host machine and paste to client machine but not the other way, what is interesting is no errors in the DEVTOOLS console.

**Reply by tricky1024** [reply]
Ok an update here, it doesnt work in mesh so i think the fix is all good for tactical but maybe it needs to be investigates directly in mesh.
Please can you test in mesh and see if that works for you.

**Reply by tricky1024** [reply]
so there is a bug reported on mesh: https://github.com/Ylianst/MeshCentral/issues/4042

**Reply by AdelantiaDev** [reply]
I already have the problem too,  never works from my Macbook to a pc, CMD+C to Ctrl+V, also didn't work from the remote pc to my Macbook, with clipboard option activated
<img width="1451" alt="Captura de pantalla 2025-02-05 a la(s) 23 46 38" src="https://github.com/user-attachments/assets/8625a1af-b454-4d8a-b34d-c21187b5ca2b" />

**Comment by zsvendo** [comment]
Even after the update, copy and paste does not work
![Captura de tela 2025-02-04 203734](https://github.com/user-attachments/assets/22e0e280-f304-4027-947c-101783662dc2)

**Reply by mtsinformatica** [reply]
It all works in both directions for me, have you checked that clipboard permissions are enabled?

**Reply by zsvendo** [reply]
Yes 
![Captura de tela 2025-02-05 172451](https://github.com/user-attachments/assets/86cc7abc-41c8-410f-b55f-9012cdcf73ce)

**Comment by Ozazjke** [comment]
I still have this issue @ v1.3.1, is there anyway to fix this?
